sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateTableExec.scala

```diff
@@ -0,0 +1,172 @@
+      throw new AnalysisException("SHOW CREATE TABLE is not supported for V2 Views.")
+    } else {
+      if (asSerde) {
+        throw new AnalysisException("SHOW CREATE TABLE is not supported for v2 tables of" +
+          " Hive Serde table to generate the Hive DDL.")
+      } else {
+        // it is used to generate Spark DDL for given table. include Hive Serde table
+        showCreateTable(table, builder)
+      }
+    }
+
+    Seq(InternalRow(UTF8String.fromString(builder.toString)))
+  }
+
+  private def showCreateTable(table: Table, builder: StringBuilder): Unit = {
+    builder ++= s"CREATE TABLE ${table.name()}"
+    val isHive = DDLUtils.isHiveTable(Option(table.properties.get(TableCatalog.PROP_PROVIDER)))
+    if (isHive) {
+      throw new AnalysisException("SHOW CREATE TABLE is not supported for Hive serde table" +
+        " to generate the Spark DDL.")
+    }
+    showTableDataColumns(table, builder)
+    showTableUsing(table, builder)
+    showTableOptions(table, builder)
+    showTableNonDataColumns(table, builder)
+    showTableComment(table, builder)
+    showTableLocation(table, builder)
+    showTableProperties(table, builder)
+  }
+
+  private def showTableDataColumns(table: Table, builder: StringBuilder): Unit = {
+    val columns = table.schema().fields.map(_.toDDL)
+    builder ++= concatByMultiLines(columns)
+  }
+
+  private def showTableUsing(table: Table, builder: StringBuilder): Unit = {
+    if (table.properties.containsKey(TableCatalog.PROP_PROVIDER)) {
+      builder ++=
+        s"USING ${escapeSingleQuotedString(table.properties().get(TableCatalog.PROP_PROVIDER))}\n"
+    }
+  }
+
+  private def showTableOptions(table: Table, builder: StringBuilder): Unit = {
+    import scala.collection.JavaConverters._
+    val dataSourceOptions = table.properties.asScala
+      .filterKeys(_.startsWith(TableCatalog.OPTION_PREFIX))
+    if (dataSourceOptions.nonEmpty) {
+      val props = dataSourceOptions.map { case (key, value) =>
+        s"'${escapeSingleQuotedString(key)}' = '${escapeSingleQuotedString(value)}'"
+      }
+
+      builder ++= "OPTIONS "
+      builder ++= concatByMultiLines(props)
+    }
+  }
+
+  private def showTableNonDataColumns(table: Table, builder: StringBuilder): Unit = {
+    if (!table.partitioning.isEmpty) {
+      val identityCols = new mutable.ArrayBuffer[String]
+      var bucketSpec = Option.empty[BucketSpec]
+
+      table.partitioning.map {
+        case IdentityTransform(FieldReference(Seq(col))) =>
+          identityCols += col
+        case t : BucketTransform =>
+          bucketSpec = Some(BucketSpec(t.numBuckets.value, t.columns.map(col => col.toString), Nil))
+        case transform =>
+          throw QueryExecutionErrors.unsupportedPartitionTransformError(transform)
+      }
+      if (identityCols.nonEmpty) {
+        builder ++= s"PARTITIONED BY ${identityCols.mkString("(", ", ", ")")}\n"
```

Comment:
In V2, it should include both the columns and transform names.

Reply:
@sunchao When we create table ,we can use follow ddl:
`CREATE TABLE t (a INT, b STRING, ts TIMESTAMP) USING parquet`
`PARTITIONED BY (`
> `a,`
> `bucket(16, b),`
> `years(ts),`
> `months(ts),`
> `days(ts),`
> `hours(ts),`
> `foo(a, "bar", 34))`

But,  `bucket(16, b),` parse to `BucketSpec` first, then convert to `BucketTransform` and add to table.partitioning when resolve plan. 

What's confusing is that we should show ddl `Partitioned BY (bucket(16, b))` or 
`CLUSTERED BY (B) [SORTED BY ()]INTO 4 BUCKETS`.  I don't really understand this place.

Reply:
@Peng-Lei yes in DSv2 bucket is just a special type of partitioning scheme, so `bucket` in either `CLUSTERED BY` or `PARTITIONED BY` will yield to the same result. Because of this, IMO I think we should convert it to `PARTITIONED BY` for the purpose of consistency.

Reply:
@sunchao Thank you for your explanation very much，I convert `BucketTransform` to `PARTITIONED BY`. Transform is great design.